src/server/middleware/openInEditorMiddleware.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright 2017-present, Callstack.
+ * All rights reserved.
+ *
+ * @flow
+ *
+ * --- OVERVIEW ---
+ *
+ * React Native apps show a stack trace when an error occurs. Each
+ * frame can be tapped on by the user. When they do, React Native will
+ * send a POST to /open-stack-frame passing along {file, lineNumber}.
+ *
+ * Haul should then open the user's editor to the file and lineNumber.
+ *
+ * What makes this gloriously fragile is, people use different editors
+ * on different platforms.
+ *
+ * We will rely on the NPM package `open-in-editor` to provide this
+ * functionality. (read: scapegoat).
+ */
+import type { $Request, $Response, Middleware } from 'express';
+import type { ReactNativeStackFrame } from '../../types';
+
+// A configuration object fed into `open-in-editor`.
+type OpenInEditorConfig = {
+  editor?: string | null,
+  cmd?: string | null,
+  pattern?: string | null,
+};
+
+const basename = require('path').basename;
+const openInEditor = require('open-in-editor');
+const editors = Object.keys(require('open-in-editor/lib/editors'));
+const messages = require('../../messages');
+const logger = require('../../logger');
+
+// does `open-in-editor` know how to detect this editor?
+const isKnownEditor = (editor: string): boolean =>
+  editors.indexOf(basename(editor)) >= 0;
+
+/**
+ * Detect the editor type to use. We use just the basename in case the user
+ * has passed a full path. Just the name work here though. Examples:
+ *
+ *   - sublime
+ *   - atom
+ *   - /usr/local/bin/atom
+ *
+ */
+function detectEditorType(): string | null {
+  // check for HAUL_EDITOR first
+  if (process.env.HAUL_EDITOR) {
+    const editor = basename(process.env.HAUL_EDITOR);
+    if (isKnownEditor(editor)) return editor;
+  }
+
+  // fallback to REACT_EDITOR
+  if (process.env.REACT_EDITOR) {
+    const editor = basename(process.env.REACT_EDITOR);
+    if (isKnownEditor(editor)) return editor;
+  }
+
+  return null;
+}
+
+/**
+ * If provided, this will be the editor command that we shell out to. You only
+ * need to use this if you've got a specific script or symlink you're trying
+ * to target.
+ */
+function detectEditorCmd(): string | null {
+  // the specific haul command gets precident
+  if (process.env.HAUL_EDITOR_CMD) return process.env.HAUL_EDITOR_CMD;
```

Comment:
Do we need `HAUL_EDITOR_CMD`? I think `HAUL_EDITOR` being either known editor or an absolute path is perfectly fine?